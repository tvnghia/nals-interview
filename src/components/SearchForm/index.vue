<template>
  <input class="form-control u-w-300 u-mr-16" type="text" v-model="valueSearch" />
</template>

<script>
import debounceSearch from '@/mixins/debounce-search'

const TIME_DEBOUNCE = 1000

export default {
  name: 'SearchForm',

  mixins: [debounceSearch],

  data() {
    return {
      valueSearch: this.$route.query.search || ''
    }
  },

  watch: {
    valueSearch: {
      handler(value) {
        // redirect to page 1 when searching
        this.handleDebounceSearch(
          () => this.$emit('on-search-blogs', { search: encodeURI(value), page: 1 }),
          value,
          TIME_DEBOUNCE
        )
      }
    }
  }
}
</script>

<style lang="scss" scoped>
.u-w-300 {
  width: 300px;
}
</style>
