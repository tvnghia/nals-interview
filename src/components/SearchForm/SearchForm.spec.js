import { shallowMount } from '@vue/test-utils'
import wrapperOps from '@/config-test/wrapper'
import SearchForm from '@/components/SearchForm'

const handleDebounceSearch = jest.spyOn(SearchForm.mixins[0].methods, 'handleDebounceSearch')

describe('SearchForm.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(SearchForm, {
      ...wrapperOps
    })
  })

  afterEach(() => {
    wrapper.vm.$destroy()
  })

  describe('when SearchForm was mounted', () => {
    it('should SearchForm render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when search blogs', () => {
    it('should call handleDebounceSearch', async () => {
      await wrapper.setData({ valueSearch: 'book' })
      expect(handleDebounceSearch).toHaveBeenCalled()
    })
  })
})
