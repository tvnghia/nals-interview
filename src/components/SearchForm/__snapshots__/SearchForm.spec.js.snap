// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm.vue when SearchForm was mounted should SearchForm render correctly 1`] = `<input type="text" class="form-control u-w-300 u-mr-16">`;
